### DISCUSSION TITLE
Local seqr - Missing HGSV (cDNA and protein) from variants

### DISCUSSION THREAD

**Discussion by gavinmonahan** [broadinstitute/seqr#2695] [Q&A]
**Describe the bug**
There are no HGSV annotations for cDNA or protein for any of the variants in our local instance of seqr, using the latest containers (also happened in previous versions too). I assume this is an issue with VEP. The only change I made to the containers is updating Hail to 0.2.77 to avoid issue #2460. Is there a way to check the output of VEP/pipeline-runner?
Also, is there a way to add in custom annotations to seqr, ie, an extra field in the INFO column?

**Link to page(s) where bug is occurring**
NA

**Scope of the bug**
It's happened for all projects we have made (and all iterations of pipeline-runner).

**Screenshots**
![image](https://user-images.githubusercontent.com/67894517/166192070-2173ef10-1517-4b6b-b3eb-efdefa67188b.png)

**Comment by hanars** [comment]
I agree that this is most likely an issue with VEP, but I would need to know which version of the docker image you are running and which command(s) you used to load the data. The latest docker image includes the correct version of hail, which means you are probably using an outdated version. In terms of checking the output, the pipeline should output an intermediate matrix table in the same directory as the input vcf with the VEP annotations. Checking if that MT has the correct VEP annotations would be a good place to start. I would recommend you check both the `vep` field in the matrix table and the `sortedTranscriptConsequences` field

**Reply by gavinmonahan** [reply]
Thanks for your help :)
I'm not sure exactly how to check, but `vep` and `sortedTranscriptConsequences` are both listed in the `metadata.json.gz` file in both the `.mt` folder and in `.mt/rows/`. `hgvsc:String` and `hgvsp:String` are also in there too. It looks like the rest of the files are binary. The folder `references` is empty.
All of the other annotations are there.
I'm using elasticsearch 7.16.3, the same version listed in the docker-compose.yml file. However, I'm assuming my issue is unrelated to that? I tried pulling elasticsearch again and it didn't download anything. That might also explain why I still had to update hail.

**Reply by hanars** [reply]
I would ask that you start by updating to the latest version of the pipeline, and make sure you have correctly downloaded all of the necessary reference data. Run `docker-compose pull pipeline-runner` to get the latest version, and then make sure you follow the initial set up instructions for loading: https://github.com/broadinstitute/seqr/blob/master/deploy/LOCAL_INSTALL.md#annotating-and-loading-vcf-callsets (in particular step 4 of Option 1 or step 3 of Option 2)

**Reply by gavinmonahan** [reply]
Update: I pulled the latest version and redownloaded the reference data, I had no issue with the pipeline but there are still no HGVScs or HGVSps. I checked the mt for those annotations and there were all "NA". I'm thinking it could be an issue with my VCFs (https://github.com/Ensembl/ensembl-vep/issues/632), but these are all generated and genotyped using GATKs best practices. 
I can try running it through VEP 99 then 106 and see if they report the HGVSs. Let me know if you have any advice
![mt no hgvs cropped](https://user-images.githubusercontent.com/67894517/167098184-f7bb0167-ac42-4327-876d-51f8990baea4.png)
Thanks for your help

**Reply by hanars** [reply]
Thanks for looking into this. VEP 99 is what comes packaged with our pipeline runner, so if you could run it through that and see if that works or not that would be helpful to determine whether there is an issue with the configuration we provide for VEP, or with that version of VEP itself. Sorry for the trouble you are having here and thanks for working with us to solve it!

**Reply by gavinmonahan** [reply]
Hi @hanars, sorry for the late reply.
I tested my VCFs twice on VEP 104 using the flags `--hgvs` or `--everything`, the problem is `--hgvs` results in exit code 1 if there is no fasta file but `--everything` simply disables the hgvs annotation steps. I checked the file `/vep_configs/vep-GRCh38-loftee.json` in the pipeline-runner container and it uses `--everything` but there was no `--fasta` flag so I added one:
`    "--fasta", "/input_vcfs/Homo_sapiens_assembly38.fasta",`
After copying the reference genome to `input_vcfs`.
It resolved the issue and I now have HGVS annotations in seqr :).
Could you please mend this in the container for future uses? I noticed there was another discussion with the same issue as well (#2241)
Thanks for your help!

**Reply by hanars** [reply]
The latest version of the docker image fixes this issue. Thanks so much for figuring this out and letting me know!

**Comment by hanars** [comment]
> Also, is there a way to add in custom annotations to seqr, ie, an extra field in the INFO column?

No, this is not supported in seqr

**Reply by gavinmonahan** [reply]
I ask because I'm interested in adding in LOD scores for big families, although I could limit the search to the linkage regions separately it would be convienient to have them on seqr so I can just filter out low LODs

**Comment by anjalijain22** [comment]
Hi @hanars! I can view the hgvs annotation for grch38 dataset but am unable to view them for Grch37 datasets. I'm using [this image](https://console.cloud.google.com/gcr/images/seqr-project/global/pipeline-runner@sha256:f2604f5935b2845e5e0fb51b22e1f150f1d128e6bc46af52fb0407d6599e87a6/details) of the pipeline runner docker image tagged as latest. The image's digest is: sha256:f2604f5935b2845e5e0fb51b22e1f150f1d128e6bc46af52fb0407d6599e87a6

Just thinking out loud here - do you think a change similar to what @gavinmonahan suggested for /vep_configs/vep-GRCh38-loftee.json would be required for the grch37 config as well?
Would appreciate your help with this!

**Reply by hanars** [reply]
please feel free to modify your config locally to see if it fixes the issue and then let me know!

**Reply by anjalijain22** [reply]
Hi @hanars,
I made changes to the VEP config locally and it works, I now have the hgvs annotations!

Similar to Gavin, I -
- Downloaded the GRCh37 reference genome to `/mnt/data/seqr/data/vep_data/homo_sapiens` directory.
- Unzipped the `Homo_sapiens.GRCh37.dna.primary_assembly.fa.gz` file
- Modified the `/mnt/data/seqr/deploy/docker/pipeline-runner/vep_configs/vep-GRCh37-loftee.json` file to add the fasta flag
`"--fasta", "/vep_data/homo_sapiens/Homo_sapiens.GRCh37.dna.primary_assembly.fa"`

Would you be able to make this fix in the pipeline runner docker image?

**Reply by hanars** [reply]
an updated docker image has been pushed up

**Comment by michael-cultrera** [comment]
Hi @hanars , just wanted to ask a question here as our hgvs annotations are missing. I looked into the matrix table and there doesn't seem to exist a `vep` field or a `sortedTranscriptConsequences` field. The --fasta flag exists in the vep config file so not too sure why annotations aren't showing. We have migrated to the new hail backend so not sure if this could be having an impact. Any help would be much appreciated ! Thanks.

**Reply by hanars** [reply]
The hail backend is strictly typed, so if you are successfully able to run searches against that backend there must be a `sorted_transcript_consequences` on your annotations hail table. I would recommend you start by looking at that field and confirming whether or not the hgvs annotations are there, and then let me know and I can see what further steps we can take to debug.

It would also be helpful if you have the logging output from one of your early loading runs from when you reloaded all the data to the hail backend, but I understand if you did not keep those around

**Reply by michael-cultrera** [reply]
Thanks for the reply @hanars ! Where exactly would I be able to find the annotations hail table?

**Reply by hanars** [reply]
In your original comment you said you checked the hail table - if you were not checking that annotations table what were you checking?

**Reply by hanars** [reply]
Can you also confirm that you have reloaded your data from scratch using one of the supported methods described here and let me know which one: https://github.com/broadinstitute/seqr/blob/master/deploy/LOCAL_INSTALL_HELM.md#using-the-load-data-page-to-load-vcf-callsets

**Reply by michael-cultrera** [reply]
Oh my mistake, I thought this was a separate table you were referring to. Doesn't have that field you are referring to.
<img width="488" alt="Screenshot 2025-04-24 at 2 25 28 PM" src="https://github.com/user-attachments/assets/82daf531-c435-4da3-b701-ae0e9d65e59e" />

**Reply by hanars** [reply]
I am not sure what mt that is, but thats not the annotations table. You should have a hail data directory with all your loaded data, it will need to be a volume mounted to your helm configuration

**Reply by hanars** [reply]
Also, per my previous comment, please confirm with me that you loaded your data using a supported method and let me know which method you used

**Reply by michael-cultrera** [reply]
We used Option 1 to load data

**Reply by michael-cultrera** [reply]
Ah yes okay seeing the annotations table with `sorted_transcript_consequences` field. All values are NA including `hgvsc: str` and `hgvsp: str`.

**Reply by bpblanken** [reply]
Hi @michael-cultrera,

It's possible that your VEP is mis-configured in some way.  Some advice on how to debug:
- Check that your VEP reference data looks as expected in `/var/seqr/vep-reference-data` and that there are valid VEP json configurations present in that directory.   
- Check that your `sorted_transcript_consequences`  is indeed missing across the whole annotations table and not just some of it (`ht.aggregate(hl.agg.collect_as_set(ht.sorted_transcript_consequences.hgvsc))` should do it!
- Try running VEP via our wrapper method on the intermediate table you posted above (the `mt.describe()`) to see if there's any logged output.  The method is present [in this file](https://github.com/broadinstitute/seqr-loading-pipelines/blob/main/v03_pipeline/lib/vep.py), and you can run it by `kubectl exec`-ing into the `pipeline-runner` pod and, in a python terminal:

```
import hail as hl

from v03_pipeline.lib.model import DatasetType, Env, ReferenceGenome
from v03_pipeline.lib.vep import run_vep

mt = hl.read_matrix_table(INTERMEDIATE_TABLE_PATH)
run_vep(mt.rows(), DatasetType.SNV_INDEL, ReferenceGenome.GRCh38)
```

**Reply by michael-cultrera** [reply]
Hi @bpblanken ,

Thanks for for getting back. I noticed that our VEP json config file `/var/seqr/vep-reference-data/GRCh38/vep-GRCh38.json` references an `/opt` directory that is empty. For example `--fasta /opt/vep/.vep/Homo_sapiens.GRCh38.dna.primary_assembly.fa.gz`, this file does not exist here, rather it exists in the same directory as the VEP json itself. Going to change this and see if it works.

**Reply by bpblanken** [reply]
Hi @michael-cultrera,

The '/opt/vep' directory is expected to be empty outside of the docker container.  That's the location VEP excepts when VEP runs _inside_ the container.  It's created by [volume mapping ](https://github.com/broadinstitute/seqr-loading-pipelines/blob/main/v03_pipeline/bin/vep#L20) `/var/seqr/vep-reference-data/GRCh38/` to  `/opt/vep/.vep/`.

**Reply by michael-cultrera** [reply]
Hey @bpblanken, makes sense thanks for clarifying. As for your previous comment regarding VEP being misconfigured:
1. VEP reference data seems to look fine inside `/var/seqr/vep-reference-data`, with there being a `vep-GRCh3{7/8}.json` within their respective reference genome folders and the json including the --fasta flags. 
2. I kept running into JVM related issues when trying to run the hail command you suggested.
3. I got logs that the struct did not have certain fields such as 

`2025-05-09 19:25:19.712 Hail: WARN: struct{amino_acids: str, biotype: str, canonical: int32, codons: str, consequence_terms: array<str>, exon: str, gene_id: str, hgvsc: str, hgvsp: str, intron: str, mane_select: str, mane_plus_clinical: str, transcript_id: str, lof: str, lof_flags: str, lof_filter: str, lof_info: str, existing_inframe_oorfs: int32, existing_outofframe_oorfs: int32, existing_uorfs: int32, fiveutr_consequence: str, fiveutr_annotation: dict<str, struct{type: str, KozakContext: str, KozakStrength: str, DistanceToCDS: str, CapDistanceToStart: str, DistanceToStop: str, Evidence: str, AltStop: str, AltStopDistanceToCDS: str, FrameWithCDS: str, StartDistanceToCDS: str, newSTOPDistanceToCDS: str, alt_type: str, alt_type_length: str, ref_StartDistanceToCDS: str, ref_type: str, ref_type_length: str}>, am_pathogenicity: float32, spliceregion: array<str>} has no field uniprot_isoform at <root>.transcript_consequences[element] for value JArray(List(JString(Q96NU1-3)))`

`2025-05-09 19:25:20.068 Hail: WARN: struct{amino_acids: str, biotype: str, canonical: int32, codons: str, consequence_terms: array<str>, exon: str, gene_id: str, hgvsc: str, hgvsp: str, intron: str, mane_select: str, mane_plus_clinical: str, transcript_id: str, lof: str, lof_flags: str, lof_filter: str, lof_info: str, existing_inframe_oorfs: int32, existing_outofframe_oorfs: int32, existing_uorfs: int32, fiveutr_consequence: str, fiveutr_annotation: dict<str, struct{type: str, KozakContext: str, KozakStrength: str, DistanceToCDS: str, CapDistanceToStart: str, DistanceToStop: str, Evidence: str, AltStop: str, AltStopDistanceToCDS: str, FrameWithCDS: str, StartDistanceToCDS: str, newSTOPDistanceToCDS: str, alt_type: str, alt_type_length: str, ref_StartDistanceToCDS: str, ref_type: str, ref_type_length: str}>, am_pathogenicity: float32, spliceregion: array<str>} has no field am_class at <root>.transcript_consequences[element] for value JString(likely_benign)`

**Reply by bpblanken** [reply]
Hi @michael-cultrera,

2. Do you have any logs for the JVM issues you encountered?
3. I'm _almost_ positive those are just warning logs that can be ignored and are safely handled by the _seqr_ implementation of VEP, which drops many fields.   If you're able to provide a few specific variants where you're seeing those logs I can try to help debug further!

**Reply by michael-cultrera** [reply]
Hi @bpblanken,

2. Tried with multiple versions of JVM, got similar output
`A fatal error has been detected by the Java Runtime Environment:
  SIGSEGV (0xb) at pc=0x000000010b2ac47f, pid=36594, tid=76807
 JRE version: OpenJDK Runtime Environment Temurin-17.0.15+6 (17.0.15+6) (build 17.0.15+6)
 Java VM: OpenJDK 64-Bit Server VM Temurin-17.0.15+6 (17.0.15+6, mixed mode, tiered, compressed oops, compressed class ptrs, g1 gc, bsd-amd64)
 Problematic frame:
 V  [libjvm.dylib+0xa8647f]  Unsafe_GetInt(JNIEnv_*, _jobject*, _jobject*, long)+0xcf
 No core dump will be written. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/http/client.py", line 287, in _read_status
    raise RemoteDisconnected("Remote end closed connection without"
http.client.RemoteDisconnected: Remote end closed connection without response`
3. Before going ahead, I just wanted to note that our directory that contains the input vcf `/mnt/seqr-data/seqr-loading-temp` does not include the intermediate matrix table (.mt folder). Hana had mentioned that this mt should be output in the same directory as the input vcf, but I'm not seeing it in that location.
